### ISSUE TITLE
image display/flickering (iTerm2 implementation issues)

### ISSUE THREAD

**Issue by klamonte** [mintty/mintty#1010]
I was testing iTerm2 image format output to mintty, and encountered the following issues:

* When the file data base64 encoding has line terminators (CR/LF), mintty emits much/all of the OSC sequence to screen.  As per the [VT320 manual](https://vt100.net/docs/vt320-uu/appendixe.html), control characters should be ignored within OSC, PM, and APC messages, and they also do not terminate the OSC string.  mintty does not have to display the image, but it should also quietly consume the entire OSC including the control characters (as xterm does).  OTHO if the newlines are intended to be stripped from the base64 data, then that needs to be documented with the image sequence because numerous [variants of base64](https://en.wikipedia.org/wiki/Base64#Variants_summary_table) require newlines.

* Images appear, but when moved/copied/destroyed leave a lot of flickering on screen where they used to be.  Flickering is behind both text and images that have covered them.  This occurs for both sixel and iTerm2 images.  But in general they do work, and so far I have not crashed mintty with sixel or iTerm2 images.

* There is no documented means for the application to detect iTerm2 image support.  mintty does not respond to DA1 with 1337 (as I have been told in some other thread I cannot remember that iTerm2 does), but it does respond with 4 (sixel).  mintty's CtrlSeqs only notes two ways to obtain its version, but no notion of what features go with what mintty versions.

* I verified that both JPG and PNG work for me.

* I verified that both ^G and ST work to close the sequences for me.  This is a difference from iTerm2's document, which only identifies ^G as the OSC terminator.

**Comment by mintty** [comment]
> When the file data base64 encoding has line terminators (CR/LF), ...

There is a limitation as mintty OSC preprocessing (which I did not touch) terminates the sequence at line ends. I considered this to be neglectable. The mintty utils repository provides an image output script command that handles this.

> As per the VT320 manual, control characters should be ignored within OSC

Please point me to the place where this is specified.

> Images appear, but when moved/copied/destroyed leave a lot of flickering on screen where they used to be.

I had once reduced the flickering potential to what I thought a minimum. What do you mean by move/copy? Using Insert/Delete Line/Column controls? Please provide a test case to demonstrate this.

> mintty does not respond to DA1 with 1337

Where is this DA1 code documented? Does it prefer to image capability or to other iTerm2 1337 OSCs as well?

**Comment by klamonte** [comment]
> > As per the VT320 manual, control characters should be ignored within OSC
> 
> Please point me to the place where this is specified.

About a third of the way down the page, under the "C1 (8-Bit) Control Characters Recognized" table, under OSC, PM, and APC:

```
Operating system command | OSC | Introduces an operating system command.*
Privacy message | PM | Introduces a privacy message string.*
Application program command | APC | Introduces an application program command.*

* The VT320 ignores all following characters, until it receives an ST control character. ESC, CAN, and SUB no longer cancel device control strings.
```

Though not authoritative, this is also consistent with Paul William's DEC parser: https://vt100.net/emu/dec_ansi_parser , in the "osc string" state.  (His parser does not account for VT52 submode's direct cursor address sequence.)

> > Images appear, but when moved/copied/destroyed leave a lot of flickering on screen where they used to be.

> I had once reduced the flickering potential to what I thought a minimum. What do you mean by move/copy? Using Insert/Delete Line/Column controls? Please provide a test case to demonstrate this.

I mean overwriting an image with more image/text, such that either some or all of the image is no longer visible.

I am testing on a Windows 7 64-bit VM running on a Linux host.  How I test:

1. Download jexer-0.3.2.jar: `wget https://sourceforge.net/projects/jexer/files/jexer/jexer-0.3.2.jar`
2. 'java -jar jexer-0.3.2.jar'  .   (For my test, I ssh from the Windows VM to the Linux host and execute it there.)
3. 'Alt-N' or click "Terminal".
4. Inside the terminal shell, 'cat cjk.txt' .  I will attach cjk.txt in the thread.
5. Open another terminal shell partially covering the CJK text, and display another image ('img2sixel something_larger.jpg').
6. Click on the CJK terminal, then the other terminal with the image that should cover it up.  You will see the CJK text "bleed through" the other image, flickering.
7. You can also do 'Open Image' from the top-left hamburger menu, select one, drag it over the terminal with CJK text, and see flickering.

If you try the same sequence in xterm, when you switch from the image to the CJK terminal underneath you may see the CJK images appear first, then the rest of the screen drawn, so there will be image-onto-image, followed by cursor positioning and text all around the images.

> > mintty does not respond to DA1 with 1337
> 
> Where is this DA1 code documented? Does it prefer to image capability or to other iTerm2 1337 OSCs as well?

I haven't seen it documented on iTerm2's site, but I recall on a thread that iTerm2 reports 1337 in DA1, which would means everything it does including images.  I could be wrong though because gnachman said he would update https://docs.google.com/spreadsheets/d/19W-lXWS9jYwqCK-LwgYo31GucPPxYVld_hVEcfpNpXg/edit#gid=907666897 in https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/11, and I don't see it in the DA1 worksheet.

Either way, I have found no documented means to detect iTerm2 image support for the application.  If you have it documented, point me to it and I will use that method.  Otherwise, I leave it to you to decide what to do (nothing is fine too if you want).

**Comment by klamonte** [comment]
[cjk.txt](https://github.com/mintty/mintty/files/4738499/cjk.txt)

For testing.

**Comment by mintty** [comment]
The OSC string thing should be fixed with my commit (noted above).

About feature detection, for specific features the usual approach is to detect the terminal type and version and set up feature usage according to knowledge about terminal support. That's what mainstream text editors also do for a variety of terminal-specific features.

For jexer, my test setup would be running it it WSL. But this jexer thing does not seem to work properly. Keyboard shortcuts are sometimes not responsive, the mouse does not work most of the time. And the cjk.txt file does not display any CJK, just a screenful of "q# ". Yet there is some flickering on the screen while the "q# " output is in the terminal window, without even any image on the screen. Therefore, my suspicion is this may be a jexer problem. I also vaguely remember that jexer would use sixel image output to display CJK, which (if true) is a really crazy and insane idea.

**Comment by mintty** [comment]
> For jexer, my test setup would be running it it WSL.

Forgot to mention, that's because if I run jexer from cygwin, it opens a new Windows console. How would I prevent it so I can run it directly in mintty?

**Comment by mintty** [comment]
About graphic CJK rendering: sorry for speaking so frankly, but I have no idea why this would be done and according to my (admittedly incomplete) analysis it seems to be the cause of the trouble here.

**Comment by klamonte** [comment]
I realize looking through the thread that my tone is a bit off.  My intention is not to demand changes or extra work, but just notify and try to help out.  I apologize if I have been coming across as hardheaded / annoying.  If this stuff was easy, it would already be done years ago. :)

> About feature detection, for specific features the usual approach is to detect the terminal type and version and set up feature usage according to knowledge about terminal support. That's what mainstream text editors also do for a variety of terminal-specific features.

I'll leave this here then.  There are multiple ideas floating around on identifying terminal capabilities, maybe one of those will catch on.  Of if mintty + iTerm2 decide together to expose iTerm2 image support in some other means then I will check it out.

> For jexer, my test setup would be running it it WSL. But this jexer thing does not seem to work properly. Keyboard shortcuts are sometimes not responsive, the mouse does not work most of the time. And the cjk.txt file does not display any CJK, just a screenful of "q# ". Yet there is some flickering on the screen while the "q# " output is in the terminal window, without even any image on the screen.

The mouse will be a WSL problem, since it works fine on many terminals (xterm, mlterm, RLogin, alacritty, vte, iTerm2, kitty, , rxvt-unicode, xfce4-terminal, xterm.js), and also mintty itself when run as the stock version in the x86-64 cygwin installer.

But there were some glitches and spotty output in jexer-0.3.2 that have since been fixed which might the cause of the not responsive keyboard shortcuts.  And while working this issue, I also found an error in Jexer that could lead to incorrect text cell pixel size calcs when telnet'd to it on a much larger than 80 x 24 screen, which could result in tiny CJK and very weird images.  That bug is now fixed in Jexer git head, but was not the root cause of the flickering for mintty.

> I also vaguely remember that jexer would use sixel image output to display CJK, which (if true) is a really crazy and insane idea.

It is a bit crazy, but there are several uses for getting good-looking output on a terminal whose actual font doesn't support some specific glyphs.  Besides CJK, the other obvious uses are powerline / nerd fonts, and glyphs for roguelike games.   But Jexer can also just emit CJK as Unicode like every other terminal does.  If one carries the technique just a bit further, you can give a multiplexed terminal window the ability to do VT100 double-width/double-height like [this](https://gitlab.com/klamonte/jexer/-/raw/master/screenshots/new_demo1.png?raw=true) and "fully pass" vttest.  (On Jexer's Xterm backend, this is technically cheating since the bigger glyphs are rendered from Jexer's terminus font and not what the terminal is using.  On its Swing backend, it's all the same font.)

That aside, whether or not applications _should_ do their own font rendering is up to them.  But it is now possible to do it in at least 5 terminals (mintty is included in that list) and 2 terminal widgets, so I would be unsurprised if it shows up in other places in the future.  Especially after vte and xterm.js wrap up and mainline their in-progress sixel support efforts.

Most important to me is terminals being able to handle whatever an application throws at them.  Even just working down this issue as a reporter has helped me fix three different bugs in two of my projects, so yay!  This has turned out to be a decent torture test for terminals with sixel support, since I have encountered only three terminals that didn't have a serious error / crash when first exposed to lots of small ephemeral images (xterm, RLogin, yaft).

I have captured some Jexer output that will hopefully make this easier to test.  When running on mintty with Lucida console 9pt font, I see some residual blinking.  On xterm with uni_vga font (-bolkhov-vga-medium-r-normal--16-160-75-75-c-80-iso10646-1, which is 8x16 pixel size) it looks like this:

![sixel_test_output](https://user-images.githubusercontent.com/4357501/83950595-8e9f8f80-a7f1-11ea-945f-eab824c4c13c.png)

The sixel output in compressed zip file:

[sixel_test.zip](https://github.com/mintty/mintty/files/4740675/sixel_test.zip)

**Comment by klamonte** [comment]
> > For jexer, my test setup would be running it it WSL.
> 
> Forgot to mention, that's because if I run jexer from cygwin, it opens a new Windows console. How would I prevent it so I can run it directly in mintty?

Try 'jexer -Djexer.Swing=false -jar jexer.jar' .  That should force the Xterm backend even on Windows.

**Comment by mintty** [comment]
> java -Djexer.Swing=false -jar jexer.jar

On cygwin 3.1.5 this gives me a screenful like ←[0;37;40m - apparently the new cygwin ConPTY gateway is not compatible with the way java/jexer sends ESC to the terminal. I'll report that to the cygwin mailing list (or you may); is there anything special how jexer outputs escape sequences?

On Cygwin 3.0.7, screen output looks OK, I see jexer showing a Demo window but it does not respond to anything (mouse or keyboard).

About sixel-test.cap, it looks actually worse on xterm than on mintty here.
How can I download the latest git commit version?

**Comment by klamonte** [comment]
> On cygwin 3.1.5 this gives me a screenful like ←[0;37;40m - apparently the new cygwin ConPTY gateway is not compatible with the way java/jexer sends ESC to the terminal. I'll report that to the cygwin mailing list (or you may); is there anything special how jexer outputs escape sequences?

Nothing special about the sequences.

My cygwin appears to be running the Windows JDK (java.exe), which does not work.  Jexer needs raw mode TTY (it spawns 'stty' to get it).

I was able to get it running against the telnet demo to avoid TTY stuff:

```
java -cp jexer.jar jexer.demos.Demo2 4444 1>/dev/null 2>/dev/null &
telnet localhost 4444
```

> About sixel-test.cap, it looks actually worse on xterm than on mintty here.
> How can I download the latest git commit version?

I put the latest jar [here](https://jexer.sourceforge.io/jexer.jar):

```
wget https://jexer.sourceforge.io/jexer.jar
```

**Comment by mintty** [comment]
If I connect to jexer via telnet, I always get a Windows cmd, even if jexer (the server) was started from cygwin. This makes testing troublesome, as a cygwin shell run from there does not behave normally. Can this be avoided?
Could you produce again, with current jexer, a terminal log like previous sixel_test.cap, that actually demonstrates the issue when catted in plain mintty?
With the current test setup, I see occasional slight flickering here and there. This is caused by regular sixel image refresh. Maybe a way can be found to handle that better.

**Comment by mintty** [comment]
I see two problems:
1. There is a mechanism in mintty to avoid refresh of already-displayed images, which does not seem to work, but also ensure that overlapped images will always display, which may not be perfect.
2. There is some explcit flicker reduction code in the case of image output but not for sixel graphics, for which the Windows bitmap stretching was considered fast enough.

Can you easily test two thing?
1. Whether the issue occurs if you assure strictly non-overlapping graphics?
2. Whether the issue occurs if you use image (iTerm2) output rather than sixel?

**Comment by klamonte** [comment]
I have captured several sixel sessions at different font sizes, and created a Jexer jar that reports the window size and text cell size.  The test inputs and outputs are [collected here](https://jexer.sourceforge.io/sixel.html), including my xterm setup.

I was also able to find two fonts for mintty under Windows that have the same dimensions as my fonts for xterm under Linux:

* Consolas 10pt (7x15).  This one had basically zero flickering under mintty with sixel, but a lot of artifacts under iTerm2.
* Courier 12pt (9x18).  This one had more flickering under sixel, but for me did not start until several seconds after the capture was finished.

> Can you easily test two things?

Sure!

> 1.  Whether the issue occurs if you assure strictly non-overlapping graphics?

I opened a single image, just display without moving it, and did not see any flicker at all.  This is the 7x15_no_overlap test on that page.

> 2. Whether the issue occurs if you use image (iTerm2) output rather than sixel?

I used mintty to create a test with iTerm2 image output, which it now on that page as the 7x15_iTerm2 test.  Obviously I do not have an xterm image for that one.  The flickering was much worse, with previously-erased images reappearing including in their original (now overwriten) locations.

The jexer-test.jar and do_test.sh will hopefully be much more useful now.  It will leave the terminal in its final state rather than restore it on exit, and should force the backend to be Xterm always.  do_test.sh also has the command shell parameter shown in a comment so you can try against things other than cmd.exe.

**Comment by mintty** [comment]
Released 3.1.8, fixing the OSC new line issue.

**Comment by mintty** [comment]
I've uploaded a fix for background and clipping artefacts of overlapped images.

**Comment by mintty** [comment]
Uploaded two fixes. Please retest.

**Comment by klamonte** [comment]
I downloaded the latest mintty-master.zip, built it in cygwin, and ran it.  Using consolas 10pt font, against the 7x15 captures at https://jexer.sourceforge.io/sixel.html .

* 'cat iTerm2_7x15.cap' -
  - Images are appearing, but getting slower as more images are added/overwritten.  Seems like a O(n^2) type operation might be happening.
  - The final screen is correct, but CPU is pegged and mintty is very slow.
  - 'clear' clears the screen (black text on grey background), CPU is still pegged.
  - 'reset' clears (grey text on black background), CPU is no longer pegged.

* 'cat sixel_7x15.cap' -
  - The first few sixel images show up, are overwritten by text, but new sixel images no longer appear.  The final screen has no images.
  - Unlike iTerm2, CPU is not pegged, but it is also not idle, looks to be about 40% of one core.
  - 'clear' clears the screen (black text on grey background), CPU is still ~40%.
  - 'reset' clears (grey text on black background), CPU is idle.

**Comment by mintty** [comment]
Images that are completely overwritten are not discarded in this version. I disabled the disposal algorithm which was buggy. Still to be analysed, but the slowness is explained for now.

Apparently it's necessary to also use the same screen resolution (DPI 100% in this case) to reproduce a screen log containing graphics; at 125% of my laptop, the result is broken by gap artefacts.
But a get a stable final result with sixel_7x15.cap as well, except that mouse mode is staill active so moving the mouse will produce escape sequences on the screen.

**Comment by klamonte** [comment]
> except that mouse mode is staill active so moving the mouse will produce escape sequences on the screen.

Yes, jexer-test.jar leaves the terminal in its application state.  Figured that might be more helpful when running the terminal under a debugger.

**Comment by mintty** [comment]
Revised detection and collection of overwritten images completely. Please retest.

**Comment by klamonte** [comment]
SUCCESS! :)

Both sixel and iTerm2 are working, no lingering flickers, final display is fully correct.  (Very quick flashing when the mouse cursor is moved around rows containing image data, which also happens on other terminals, but I don't consider that a bug.)

Sixel performance is a bit faster than iTerm2, but overall both are very usable and look great.

From my perspective, this issue is solved.

**Comment by mintty** [comment]
Thanks for testing. Actually, just a few minutes ago I fixed another bug which would leave artefacts in the padding area (top pixel row) when scrolling in the scrollback buffer... I had just removed that code because its purpose was unclear; now documented. I guess I'll make that a release.

**Comment by mintty** [comment]
Released 3.2.0.

**Comment by traut** [comment]
I'm experiencing a similar issue – wild flickering in the iTerm _if the hyperlink output is used_. Not only that, CPU usage of iTerm2 jumps to 70-80%. Setting `refresh_per_second` to 5 seconds helped a bit when only a few progress bars are shown, but has no effect on flickering with, say, 20 progress bars active
